src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java

```diff
@@ -63,28 +63,37 @@
   }
 
   ManagedChannel getChannel() {
-    ManagedChannel mc = channelRef.get();
-    if (mc == null) {
+    ManagedChannel managedChannel = channelRef.get();
+    if (managedChannel == null) {
       synchronized (channelRef) {
-        mc = channelRef.get();
-        if (mc == null) {
-          NameResolver.Factory resolverFactory = builder.getNameResolverFactory();
-          if (resolverFactory == null) {
-            resolverFactory = SimpleNameResolverFactory.forEndpoints(builder.getEndpoints());
-          }
-
+        managedChannel = channelRef.get();
+        if (managedChannel == null) {
           ManagedChannelBuilder<?> channelBuilder = builder.getChannelBuilder();
+
+          // If channel builder is provided, any additional configuration such as SslContext, name
+          // resolver or load balancer is ignored
           if (channelBuilder == null) {
-            channelBuilder = defaultChannelBuilder(resolverFactory);
+
+            NameResolver.Factory resolverFactory = builder.getNameResolverFactory();
+            if (resolverFactory == null) {
+              resolverFactory = SimpleNameResolverFactory.forEndpoints(builder.getEndpoints());
+            }
+
+            channelBuilder = defaultChannelBuilder();
+            channelBuilder.nameResolverFactory(resolverFactory);
+
+            if (builder.getLoadBalancerFactory() != null) {
+              channelBuilder.loadBalancerFactory(builder.getLoadBalancerFactory());
```

Comment:
if`LoadBalancerFactory` is not given, what's the default grpc behavior on choosing endpoints?